[1단계 - 로또 구현] 후유(김대동) 미션 제출합니다.

eoehd1ek opened the pull request:
안녕하세요. 우테코 BE 7기 후유입니다.
코드 리뷰 잘 부탁드립니다.

아직 작성한 코드에서 아쉬운 점이나 미숙한 점을 스스로 찾기 힘들어서 질문을 많이 작성하지 않은 점 양해 부탁드립니다.
개선해야 할 코드가 있다면 해결 방법을 바로 알려주시는 것도 좋겠지만,
"~과 같은 상황이 발생하면 어떻게 될까요?" 처럼 퀴즈를 주셔서 제 스스로 코드 개선점을 생각해 보는 기회도 마구마구 왕왕 주셔도 좋습니다.

미션1 로또 진행 이후 프롤로그 회고글: https://prolog.techcourse.co.kr/studylogs/3829

---

LottoController.run 메소드는 다음과 같은데요.
```java
public void run() throws IOException {
        Money money = inputView.inputMoney();
        final int lottoCount = Lotto.getLottoCount(money);
        outputView.printLottoCount(lottoCount);

        List<Lotto> lottos = lottoManager.generateLottos(lottoCount);
        outputView.printLottos(lottos);

        WinningLotto winningLotto = inputView.inputWinningLotto();

        WinningResult winningResult = lottoManager.getWinningResult(lottos, winningLotto);
        outputView.printWinningResult(winningResult);
        outputView.printRevenue(lottoManager.getRevenue(winningResult, money));
    }
```
위 run() 메소드 본문이 커지게 된다면 각각의 부분을 LottoController 내부에 메소드로 분리하고, run() 메소드에서 호출할 것 같은데요.
그렇게 되면 메소드 내부에서 사용하는 Money, lottoCount, List<Lotto> lottos와 같은 변수는 어떻게 관리해야 할지 궁금합니다.
당장 제가 구현하게 된다면 LottoController 전역 변수에 Money, lottoCount를 작성할 것 같습니다.
그러면 run() 메소드가 2번 이상 호출 되는 경우 문제가 발생할 여지가 있을 것 같습니다.

eoehd1ek committed:
docs: 기능 구현 목록 정리

eoehd1ek committed:
feat: 구입 금액 입력 기능 추가

eoehd1ek committed:
feat(Lotto): 로또 번호 클래스 추가

eoehd1ek committed:
feat(RandomNumbersGenerator): 난수 생성 유틸리티 클래스 추가

eoehd1ek committed:
feat(*): 로또 구입 기능 추가

eoehd1ek committed:
refactor(*): 구입 금액 입력 리팩토링

eoehd1ek committed:
feat(*): 당첨 번호와 보너스 번호 입력 기능 추가

eoehd1ek committed:
feat(WinningLotto): 당첨 로또 클래스 구현

eoehd1ek committed:
feat(WinningInfo): 등수별 상금 값을 구할 수 있는 enum 추가

eoehd1ek committed:
feat(WinningInfo): 번호 매칭 개수 따라서 순위를 계산하는 기능 추가

eoehd1ek committed:
feat(LottoManager): 순위별 당첨 횟수를 계산하는 기능 추가

eoehd1ek committed:
refactor: 로또 숫자 타입을 Number 객체로 변경

eoehd1ek committed:
fix(Lotto): 출력 형식 오류 수정

로또 번호를 vo 객체로 변경 하면서 getter를 누락하여 출력 양식이 잘못된 버그 수정

eoehd1ek committed:
feat(*): 로또 당첨 개수 목록 출력 기능 추가

eoehd1ek committed:
feat(*): 수익률 계산 기능 추가

eoehd1ek committed:
feat(LottoController): 수익률 출력 기능 추가

eoehd1ek committed:
refactor(LottoManagerTest): 테스트 내부 중복 코드 개선

eoehd1ek committed:
refactor(LottoController): 입출력 책임을 view 계층에 위임하도록 수정

eoehd1ek committed:
refactor: 컨트롤러의 메소드를 다른 객체의 책임으로 분리

eoehd1ek committed:
fix(Lotto): 로또 가격 생성자 인자 오류 수정

eoehd1ek committed:
refactor(WinningLotto): 코드 depth 제한에 맞게 수정

eoehd1ek committed:
refactor: Money의 문자열 정적 생성자 추가

eoehd1ek committed:
chore(WinningLotto): 미사용 메소드 삭제

eoehd1ek committed:
feat(Lotto): 로또 번호를 정렬된 상태로 저장

EastHShin commented on src/main/java/Lotto.java:
```diff
@@ -0,0 +1,59 @@
+                .toList();
+    }
+
+    public static int getLottoCount(Money money) {
```
get 이라는 메서드의 네이밍은 단순히 값을 꺼낸다는 의미로 받아들일 수 있을 것 같은데, 해당 메서드는 로또 개수의 계산을 담당하고 있네요! 명확히 표현해보면 어떨까요?

eoehd1ek replied:
계산되는 값을 얻는 메서드는 count, sum, average, size, location, calculate, diff 과 같은 명사로 시작되어도 괜찮겠네요. 
전달받은 Money에서 구매 가능한 로또의 개수를 센다는 의미로
countPurchasableLottosByMoney 로 변경했습니다.
생각해보게 해 주셔서 감사합니다.

EastHShin commented on src/main/java/Lotto.java:
```diff
@@ -0,0 +1,59 @@
+        return numbers.stream().map(Number::getValue).toList();
+    }
+
+    public String getInfo() {
```
Lotto 라는 도메인 클래스에서 출력 형식에 대한 책임을 갖고 있네요!
출력 형식이 도메인 모델에서 알고 있어야 하는 정보일지 한 번 생각해보시면 좋을 것 같아요!

eoehd1ek replied:
출력 형식은 자주 변경될 수 있고, 이에 출력 형식이 변경될 때 마다 도메인 모델이 변경되는 문제가 발생할 수 있겠네요.
출력 형식에 대한 책임은 출력을 담당하는 view에 있어야 할 것 같습니다.

view는 출력 형식에 맞게 전달 받은 객체에서 값을 꺼내 출력하도록 코드를 변경하겠습니다.

감사합니다.

EastHShin commented on src/main/java/WinningLotto.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    private void validate(Lotto winningNumbers, Number bonusNumber) {
+        if (winningNumbers.getNumbers().stream().anyMatch(number -> number == bonusNumber.getValue())) {
```
if 문 내부 로직을 적절한 메서드 분리를 통해 명확히 할 수 있을 것 같아요!

eoehd1ek replied:
스트림을 사용한 코드에서 한 줄에 하나의 '.' 만 가지도록 변경했습니다.
변경 후 이전보다 가독성이 훨씬 좋아졌네요.
앞으로는 메소드 체이닝을 사용할 때 한 줄에 하나의 '.'만 가지도록 하겠습니다.
감사합니다.

EastHShin commented on src/test/java/LottoManagerTest.java:
```diff
@@ -0,0 +1,59 @@
+
+    @DisplayName("로또와 당첨 번호를 토대로 로또 순위별 당첨 횟수를 알 수있다.")
+    @Test
+    void test1() {
```
테스트 메서드명을 명확하게 명시하는 건 어떨까요? DisplayName과 테스트 메서드명은 어떤 역할의 차이를 가지고 있을까요?!

eoehd1ek replied:
@DisplayName를 사용하면 IDE에서 해당 내용이 표기됩니다. 그래서 다음과 같이 이 테스트에서 어떤 작업을 했는지, 이 테스트로 무엇을 확인하고 싶은지 작성한다면
어떤 동작의 테스트가 실패했는지 쉽게 인지할 수 있을 것 같습니다.
```
~작업을 할 때 ~인지 검사한다
~작업을 할 때 ~하다면 예외가 발생해야 한다
```

테스트 메서드명에는 어떤 동작을 테스트 하는지 작성해야 한다고 생각합니다.

좋은 질문 감사합니다.

EastHShin replied:
네ㅎㅎ 테스트 코드는 어떤 기능을 가지고 있는지에 대한 명세의 역할도 한다고 생각해요. 그런 관점으로 테스트 메서드 명을 명확히 적어주면 좋다고 생각합니다!

EastHShin commented on src/test/java/LottoTest.java:
```diff
@@ -0,0 +1,64 @@
+    @Test
+    void test4() {
+        // given
+        List<Integer> numbers = List.of(0, 1, 2, 3, 4, 46);
+
+        // when & then
+        assertThatThrownBy(() -> new Lotto(numbers))
+                .isInstanceOf(IllegalArgumentException.class);
```
현재 이 테스트 케이스로는 0에서 예외가 발생한건지, 46에서 예외가 발생한건지 특정이 어려워보여요!
두 케이스를 모두 포함한 테스트를 만드려면 어떻게 해야할까요?

eoehd1ek replied:
@ParameterizedTest 를 활용해
예외 범위인 0을 포함한 케이스와 46을 포함한 케이스를 각각 테스트 하도록변경했습니다.
감사합니다.

eoehd1ek replied:
@MethodSource 어노테이션에서 사용하는 메서드의 이름 규칙에 질문이 있습니다.

@MethodSource 를 공부하며 본 예제들에서 다음과 같은 이름 규칙을 봤는데요.
```
반환값+Provider
반환값+Arguments
```
이번엔 지금까지 동사+명사 형식의 메소드 이름으로 작성해서 generateExceptionNumber 으로 메소드 이름을 정했습니다.

그런데 앞으로는 아래 이유로 반환값+Arguments 형식으로 작성할 것 같습니다.
1. 테스트 메서드 안에서는 @MethodSource 어노테이션에서 사용하는 정적 메소드를 사용하는 경우는 없을 것 같음.
2. 매개변수의 의미를 포함하고 있음.
3. @MethodSource 어노테이션에서 사용한다는 느낌을 주기 위해서 동사+명사 형식 대신 접미어 Arguments으로 끝냄.

혹시 위 생각이 괜찮을지 리뷰어님의 의견 받을 수 있을까요?

EastHShin replied:
네 저도 괜찮다고 생각되어요! 이러한 부분은 정답이 없으니 후유만의 기준을 잘 정립해 사용하는 이유를 말씀하실 수 있으면 됩니다!
정답이 없는 문제는 보통 동료들과의 협의를 통해 이루어지니 이 때, 후유만의 기준이 잘 정립되어 있다면 큰 도움이 될거에요 😄

EastHShin commented on README.md:
```diff
@@ -1,3 +1,76 @@
+    - 콤마 및 공백으로 구분된 6개의 숫자열 입력 
+
+## 기능 요구사항
+- [x] 로또 구입 금액을 입력하면 구입 금액에 해당하는 로또를 발급해야 한다.
```
todo 리스트를 통하 기능명세 👍

EastHShin commented on src/test/java/WinningLottoTest.java:
```diff
@@ -0,0 +1,56 @@
+        // then
+        assertThat(matched).isEqualTo(expected);
+    }
+}
```
> 위 run() 메소드 본문이 커지게 된다면 각각의 부분을 LottoController 내부에 메소드로 분리하고, run() 메소드에서 호출할 것 같은데요.
그렇게 되면 메소드 내부에서 사용하는 Money, lottoCount, List lottos와 같은 변수는 어떻게 관리해야 할지 궁금합니다.
당장 제가 구현하게 된다면 LottoController 전역 변수에 Money, lottoCount를 작성할 것 같습니다.
그러면 run() 메소드가 2번 이상 호출 되는 경우 문제가 발생할 여지가 있을 것 같습니다.

질문 주신것에 대한 답변을 드리자면,
상태를 가지는 경우에 대한 고민으로 보입니다! run 이 두번 호출된다는 것은, 별개의 로또 게임을 수행한다고 봐도 되는 것이고, 이는 곧 새로운 컨트롤러를 생성하거나 컨트롤러 내부의 상태를 초기화 한 뒤 게임을 실행해야 하는 조건으로 받아들일 수 있을것 같습니다. 그렇다면 말씀주신 부분이 해결이 될 것 같아요.

그런데 다시 돌아가서, run 이 커질경우에 왜 전역변수로 할당해야 하는지 생각해볼 필요가 있어보여요! 적절한 단위의 메서드로 분리하고, 지역변수를 넘겨주는 방식으로도 충분히 해결 가능해보입니다.

eoehd1ek replied:
한번 생성한 컨트롤러는 프로그램이 종료되기 전 까지 재사용 된다는 생각에 run() 메소드도 여러 번 호출 될 것이라 생각했습니다. 그래서 run() 메소드로 생성된 이전 데이터로 문제가 발생하지 않을까 했는데, 정말로 새로운 컨트롤러를 생성하거나 컨트롤러 내부 상태를 초기화 하는 방법을 사용하면 되겠네요.
답변 감사합니다.

EastHShin reviewed (approved):
안녕하세요 후유!
이번 미션 함께 진행하게 된 이스트입니다 😄 

첫 미션임에도 기능 잘 구현해주셨네요!

코멘트와 질문에 대한 답변 남겨 두었으니, 확인해주시고 질문이 있으시다면 코멘트나 DM 주시면 답변 드리겠습니다!

EastHShin reviewed (changes requested):
실수로 approve를 해버렸네요 😢 
위 코멘트를 바탕으로 수정 후 다시 리뷰요청 주시면 됩니다!

eoehd1ek committed:
refactor(Lotto): 구매 가능한 로또 수를 얻는 메서드 이름을 명확하게 변경

getLottoCount -> countPurchasableLottosByMoney

eoehd1ek committed:
style: 스트림 사용에서 한 줄에 하나의 '.' 만 가지도록 변경

eoehd1ek committed:
test(Lotto): 로또 범위가 벗어나는 경우 예외 발생 테스트 보완

기존 방식에서는 0과 46 중 어떤 상황에서 예외가 발생했는지 확인하기 어려움.

이에 0이 포함된 경우와 46이 포함된 경우 2가지 케이스를 각각 검사하도록 함.

eoehd1ek committed:
refactor(LottoManagerTest): 명확한 테스트 의도를 표현하도록 메소드 이름 수정

eoehd1ek committed:
refactor(LottoTest): 명확한 테스트 의도를 표현하도록 메소드 이름 수정

eoehd1ek committed:
refactor(MoneyTest): 명확한 테스트 의도를 표현하도록 메소드 이름 수정

eoehd1ek committed:
refactor(RandomNumbersGeneratorTest): 명확한 테스트 의도를 표현하도록 메소드 이름 수정

eoehd1ek committed:
refactor(WinningInfoTest): 명확한 테스트 의도를 표현하도록 메소드 이름 수정

eoehd1ek committed:
refactor(WinningLottoTest): 명확한 테스트 의도를 표현하도록 메소드 수정

eoehd1ek committed:
fix(RandomNumbersGeneratorTest): 올바르지 못한 로직으로 실패하는 테스트 수정

eoehd1ek committed:
refector(Lotto, outputView): 도메인 모델이 출력 형식을 가지지 않도록 수정

eoehd1ek committed:
refector(WinningInfo, outputView): 도메인 모델이 출력 형식을 가지지 않도록 수정

eoehd1ek committed:
refector(outputView): 한 단계의 들여쓰기만 하도록 수정

eoehd1ek committed:
refactor(LottoManager): 수익률을 계산하는 메서드 이름을 명확하게 변경

eoehd1ek committed:
refactor(LottoManager): 우승 결과를 계산해 반환하는 메서드의 이름을 명확하게 변경

eoehd1ek committed:
style: 한 줄에 하나의 '.' 만 사용하도록 코드 스타일 변경

eoehd1ek committed:
test(LottoTest): 누락된 구매 가능한 로또 개수 반환 메서드 테스트 추가

wad-shindonghoon reviewed (approved):
안녕하세요 후유!

수정해주신 부분 확인했습니다 👏 
2단계 미션에서 뵈어요!

EastHShin commented on src/main/java/LottoController.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.List;
+
+public class LottoController {
+    private InputView inputView;
```
<img width="585" alt="image" src="https://github.com/user-attachments/assets/e073b870-b8d3-438c-922a-82deb80dbedb" />

IDE 에서 final을 붙이는게 어떻냐고 제안을 하네요! 왜 이런 제안을 해주었을지 알아보아요!

eoehd1ek replied:
final을 붙이면 해당 변수는 초기화 이후 재할당을 제한합니다.
따라서 멤버 상수, 값이 변함을 원치 않는 대상에 final을 사용한다면 의도치 않은 값 변경을 예방할 수 있을 것 같습니다.
감사합니다.

EastHShin reviewed (approved):
안녕하세요 후유!

수정해주신 부분 확인했습니다 👏 
2단계 미션에서 뵈어요!
